[CALCITE-2460][CALCITE-2459] Add implementation of To_Base64 and  From_Base64 to SqlFunctions

wenhuitang opened the pull request:
This PR for issue https://issues.apache.org/jira/browse/CALCITE-2460 and https://issues.apache.org/jira/browse/CALCITE-2459 adds implementation of To_Base64 and From_Base64 to SqlFunctions

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -107,6 +109,18 @@
+      return null;
+    }
+    try {
+      return Base64.encodeBase64String(s.getBytes("UTF-8"));
```
Calcite is Java8. Please use https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html

wenhuitang replied:
Thanks. I have changed it.

yanghua commented:
Hi @vlsi and @julianhyde , Any opinion and comment about this PR?

asereda-gs commented:
Hi @wenhuitang I'll try to review this PR.

Meanwhile, can you please add your name to commit message in parenthesis

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -107,6 +108,35 @@
+    }
+    try {
+      return Base64.getEncoder().encodeToString(s.getBytes("UTF-8"));
+    } catch (Exception e) {
```
Do we expect an exception here ?

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -107,6 +108,35 @@
+        builder.append((char) b);
+      }
+      return builder.toString();
+    } catch (Exception e) {
```
[decode](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Decoder.html#decode-java.lang.String-) throws `IllegalArgumentException` maybe catch it instead of generic exception ?

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -107,6 +108,35 @@
+    try {
+      StringBuilder builder = new StringBuilder();
+      byte[] afterDecode = Base64.getDecoder().decode(s);
+      for (byte b : afterDecode) {
```
why not use `new String(...)` instead of loop ?

asereda-gs commented on core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:
```diff
@@ -5171,6 +5200,40 @@ private void checkNullOperand(SqlTester tester, String op) {
+  }
+
+  @Test public void testFromBase64() {
+    tester.setFor(MySqlOperatorTable.FROM_BASE64);
```
Is there a test for `from_base64` with invalid base64 string (not empty but incorrect) ?

asereda-gs replied:
Also there are some [MySQL specifics](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_to-base64) for base64:

* A newline is added after each 76 characters of encoded output to divide long output into multiple lines.
* Decoding recognizes and ignores newline, carriage return, tab, and space.

Should it exactly replicate same functionality ?

asereda-gs commented:
Also can you please rebase against latest master ? I want to make sure build doesn't fail.

wenhuitang commented:
> Also can you please rebase against latest master ? I want to make sure build doesn't fail.

Hello, @asereda-gs  thanks for you review, I made some changes according to your comments and rebase against latest master. Do you have time to have a look at the PR again?

asereda-gs commented:
@wenhuitang I will check.

Please add you name to commit message. It should say:
> [CALCITE-2460] [CALCITE-2459] Add TO_BASE64 / FROM_BASE64 SQL functions (FirstName LastName)

asereda-gs commented on src/main/config/checkstyle/checker.xml:
```diff
@@ -95,7 +95,7 @@ limitations under the License.
     <module name="EqualsHashCode"/>
       <!-- Disallow unnecessary instantiation of Boolean, String -->
     <module name="IllegalInstantiation">
-      <property name="classes" value="java.lang.Boolean, java.lang.String"/>
```
Why are we changing this ?

wenhuitang replied:
Ok， I use a loop to convert byte[] to String instead of changing the checker.xml.

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -121,6 +124,36 @@
+    if (string == null) {
+      return null;
+    }
+    String base64 = new String(Base64.getEncoder().encode(string.getBytes(UTF_8)), UTF_8);
```
you can use [Encoder.encodeToString()](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.Encoder.html#encodeToString-byte:A-) function

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -121,6 +124,36 @@
+    }
+    String base64 = new String(Base64.getEncoder().encode(string.getBytes(UTF_8)), UTF_8);
+    String temp = base64;
+    StringBuilder str = new StringBuilder();
```
probably can be initialized at capacity `base64.length() + base64.length() / 76`

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -121,6 +124,36 @@
+    String base64 = new String(Base64.getEncoder().encode(string.getBytes(UTF_8)), UTF_8);
+    String temp = base64;
+    StringBuilder str = new StringBuilder();
+    while ((temp.length() / 76) > 0) {
```
Does it make sense to use [Mime decoder](https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html#mime) directly ? 

Can you check if MySQL produces mime version of Base64 encoding ?

wenhuitang replied:
> MIME
Uses the "The Base64 Alphabet" as specified in Table 1 of RFC 2045 for encoding and decoding operation. The encoded output must be represented in lines of no more than 76 characters each and uses a carriage return '\r' followed immediately by a linefeed '\n' as the line separator. No line separator is added to the end of the encoded output. All line separators or other characters not found in the base64 alphabet table are ignored in decoding operation.

MySQL specifies that "A newline is added after each 76 characters of encoded output to divide long output into multiple lines." I haved checked the binary result of TO_BASE64 ,it seems that it not uses carriage return.

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -121,6 +124,36 @@
+      return null;
+    }
+    try {
+      base64 = base64.replaceAll("[\\t\\n\\r ]", "");
```
Probably `\s` is better:

From docs 
> \s  A whitespace character: [ \t\n\x0B\f\r]

asereda-gs commented:
Also pls address Julian's comments in JIRA(eg BINARY vs VARCHAR)

wenhuitang commented:
> Also pls address Julian's comments in JIRA(eg BINARY vs VARCHAR)

Ok, I'll address it ,Thanks.

wenhuitang commented:
> Also pls address Julian's comments in JIRA(eg BINARY vs VARCHAR)

I have addressed some comments. But as for "TO_BASE64 takes a BINARY or VARBINARY, not a VARCHAR, as argument." , considering Mysql specifies that "Converts the string argument to base-64 encoded form and returns the result as a character string with the connection character set and collation. " https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_to-base64.  I'm not sure if it must change its argument to BINARY or VARBINARY?  So I let TO_BASE64 can take both string and  binary string as argument.  Hope get some advises. Thanks.

wenhuitang commented:
@asereda-gs @julianhyde  Hello， I have made some modification, welcome comment. Thanks.

asereda-gs commented:
@wenhuitang I will check over week-end

asereda-gs commented on site/_docs/reference.md:
```diff
@@ -2104,6 +2104,11 @@ semantics.
 | m | JSON_STORAGE_SIZE(jsonValue)                   | Returns the number of bytes used to store the binary representation of a *jsonValue*
 | o | LEAST(expr [, expr ]* )                        | Returns the least of the expressions
 | m p | LEFT(string, length)                         | Returns the leftmost *length* characters from the *string*
+| m | REVERSE(string)                                | Returns the reverse order of *string*
```
Is this documentation from different commit ? 
I think this one only adds `TO_BASE64` / `FROM_BASE64`

asereda-gs commented:
Overall LGTM. Please address small comments (in docs) and squash commits into one.

wenhuitang commented:
> Overall LGTM. Please address small comments (in docs) and squash commits into one.

Hi @asereda-gs ，I have already addressed. Thanks a lot for your time on this PR.

wenhuitang commented:
> Rather then `(temp.length() / 76) > 0` why not write `temp.length() > 76`?
> 
> Punctuation consistent with surrounding code. Rather than `/** SQL TO_BASE64(string) function*/` write `/** SQL TO_BASE64(string) function. */`
> 
> Rather than `assertEquals(expected, actual)` use `assertThat(actual, is(expected))`. Easier to read.
> 
> Your algorithm creates a temp string and chews it down shorter and shorter. That is an O n^2 algorithm. Instead create one string, an offset into it (`for (int i = 0; i + 76 < s.length(); i += 76)`) and use `StringBuilder.append(s, i, i + 76)` or similar.

Hello, @julianhyde Thanks for your time on this PR, I have addressed the comments,  do you have time to have a look at the PR again?

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +126,48 @@
+    StringBuilder str = new StringBuilder(base64.length() + base64.length() / 76);
+    int i;
+    if (base64.length() > 76) {
+      for (i = 0; i + 76 < base64.length(); i += 76) {
```
It looks there is a bit of code duplication between `toBase64(String string)` and `toBase64(ByteString)` methods.

Can you use Guava's [Splitter.fixedLength()](https://guava.dev/releases/23.0/api/docs/com/google/common/base/Splitter.html#fixedLength-int-) here ?

asereda-gs commented:
@wenhuitang thanks for addressing Julians comments.
I left one more. Pls take a look

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +126,48 @@
+  /** SQL FROM_BASE64(string) function. */
+  public static ByteString fromBase64(String base64) {
+    try {
+      base64 = base64.replaceAll("[\\t\\n\\r\\s]", "");
```
Please cache the regexp in question.

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +126,48 @@
+  public static ByteString fromBase64(String base64) {
+    try {
+      base64 = base64.replaceAll("[\\t\\n\\r\\s]", "");
+      return new ByteString(Base64.getDecoder().decode(base64.getBytes(UTF_8)));
```
This looks a bit strange.  Why do you use `base64.getBytes(UTF_8)` here?
Isn't `Base64.getDecoder().decode(base64)` enough?

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +126,48 @@
+        str.append(base64, i, i + 76);
+        str.append("\n");
+      }
+      str.append(base64, i, base64.length());
```
Does this work for empty string?

Can you please add a test with `to_base64('')` ?

wenhuitang commented:
Hi, @asereda-gs @vlsi, thanks a lot, I have already addressed the new comments.

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +129,39 @@
+  /** SQL FROM_BASE64(string) function. */
+  public static ByteString fromBase64(String base64) {
+    try {
+      base64 = FROM_BASE64_REGEXP.matcher(base64).replaceAll("");
```
Just wondering: could you please explain why `java.util.Base64#getMimeDecoder` does not work here?
It looks like `java.util.Base64#getMimeDecoder` is just enough, and it does not require regexp preprocessing

wenhuitang replied:
When I using Base64.getMimeDecoder().decode(base64), the result of from_base64('-100') is d74d, however it should be null in MySql.

vlsi commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +129,39 @@
+    String base64 = Base64.getEncoder().encodeToString(bytes);
+    StringBuilder str = new StringBuilder(base64.length() + base64.length() / 76);
+    if (base64.length() > 76) {
+      Splitter.fixedLength(76).split(base64).iterator().forEachRemaining(s -> {
```
Can you please clarify why do you split the lines manually instead of using `java.util.Base64#getMimeEncoder()` ?
It looks like `Base64` can split the lines on its own just fine.

wenhuitang replied:
> MIME
Uses the "The Base64 Alphabet" as specified in Table 1 of RFC 2045 for encoding and decoding operation. The encoded output must be represented in lines of no more than 76 characters each and uses a carriage return '\r' followed immediately by a linefeed '\n' as the line separator. No line separator is added to the end of the encoded output. All line separators or other characters not found in the base64 alphabet table are ignored in decoding operation.

I did it because I didn't find any evidence that MySql produces mime version of Base64 encoding and MySQL specifies that 
"A newline is added after each 76 characters of encoded output to divide long output into multiple lines." 
"Decoding recognizes and ignores newline, carriage return, tab, and space." 
I haved checked the binary result of TO_BASE64 in MySql , it seems that it not uses carriage return followed immediately by a linefeed as the line separator.

asereda-gs commented on core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:
```diff
@@ -4277,6 +4277,76 @@ private void checkNullOperand(SqlTester tester, String op) {
   }
 
+  @Test public void testToBase64() {
+    final SqlTester tester1 = tester(SqlLibrary.MYSQL);
```
can you pls add test for empty / single char string(s) ?

wenhuitang replied:
I have addressed. Thanks ^^.

asereda-gs commented:
@vlsi / @julianhyde I'm fine with this review. 
Do you have any concerns if I merge it ?

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -123,6 +129,39 @@
+  private static String toBase64_(byte[] bytes) {
+    String base64 = Base64.getEncoder().encodeToString(bytes);
+    StringBuilder str = new StringBuilder(base64.length() + base64.length() / 76);
+    if (base64.length() > 76) {
```
Do you really need this check ? Splitter will probably create single element iterator anyway.

wenhuitang replied:
sorry，I didn't notice that，I have already addressed.

asereda-gs commented on core/src/test/java/org/apache/calcite/test/SqlFunctionsTest.java:
```diff
@@ -76,6 +78,34 @@
     assertEquals("a bcd iijk", lower("A bCd Iijk"));
   }
 
+  @Test public void testFromBase64() {
```
Can you also add to/from tests ?  Presumably `fromBase64(toBase64(x)) == x`. 
```java
// add more examples as you see fitted
for(String expected: Arrays.asList("", "0", "a", " ", "\n", "\r\n", "\u03C0")) {
  String result = fromBase64(toBase64(code));
  assertEquals(result, expected);
}
```

wenhuitang replied:
I have added the to/from tests.
